### PR TITLE
feat: Physics-based optimal P estimation with prop-size-aware recommendations

### PR DESCRIPTION
**Pull request by nerdCopter**
test-binaries. commit `84a56e2` 
[binary-windows.zip](https://github.com/user-attachments/files/25772794/binary-windows.zip)
[binary-macos.zip](https://github.com/user-attachments/files/25772795/binary-macos.zip)
[binary-linux.zip](https://github.com/user-attachments/files/25772796/binary-linux.zip)

---

TL;DR: "generates/analyzes step response, then compares that to physics-based targets per prop-size. targets are empirical values that need statistical research still."

---

## Overview

Adds physics-aware optimal P gain estimation based on Brian White's (PIDtoolbox author @bw1129) theory: **optimal response timing is aircraft-specific**, determined by power-to-rotational-inertia ratio.

## What This Does

Analyzes your flight logs and recommends optimal P gains (and corresponding D values) based on:
- **Response time (Td)**: How quickly your quad reaches 50% of setpoint
- **Noise levels**: Detects if you're hitting thermal/mechanical limits
- **Consistency**: Warns if measurements are unreliable

### Basic Usage (generates step response plots only):
```bash
./target/release/BlackBox_CSV_Render ./input/*.csv --step --estimate-optimal-p --prop-size 5 --output-dir ./output
```

**Important:** Use `--step` flag to skip other graphs (faster testing). The feature only affects step response analysis.

### Prop Size Selection:
- Match your **actual propeller diameter**, not frame size
- Example: 6" frame with 5" props → use `--prop-size 5`
- Example: 7" frame with 6" props → use `--prop-size 6`
- Supported: 1-15 inches

#### Console Output:
```
Roll: Td=18.3ms (target 20.0±5.0ms, -8% dev), Noise=LOW, Consistency=79%
  Current P=53
    → Increase recommended:
      Conservative: P≈54 (+1), D≈45 (+3)
    Response is faster than target with low noise levels...
```

#### PNG Legend (appears on step response plots):
```
Optimal P (5")
  Td: 18.3ms (target: 20.0ms)
  Deviation: -8% (WITHIN TARGET)
  Noise: LOW
  Recommendation: P≈54 (+1), D≈45 (+3)
```

#### Warning Example:
```
[WARNING] High variability (CV=44%) - results may be unreliable
```
This means your stick inputs were inconsistent. Fly cleaner flips/rolls for better measurements.

## Key Features

1. **Prop-size-aware targets** (1"-13" props supported)
   - Each prop size has physics-based response time targets
   - Automatically scales: smaller props = faster response, larger props = slower

2. **Complete tuning guidance**
   - Recommends both P **and** D values together
   - Uses recommended P:D ratio from step response analysis
   - Conservative tier only

3. **Noise analysis**
   - LOW noise = headroom to increase P
   - HIGH noise = at thermal/mechanical limits
   - Helps avoid motor overheating

4. **Quality control**
   - Warns if measurements are unreliable
   - Shows consistency metrics
   - Helps identify mechanical issues

## What to Test

1. **Accuracy**: Do recommendations make sense for your quad?
2. **Prop size logic**: Test with different prop sizes (if you have multiple quads)
3. **Edge cases**: Very old logs, missing data, unusual configurations
4. **Clarity**: Are recommendations easy to understand?
5. **PNG readability**: Is the legend information useful?

## Changes vs. Main

- **No breaking changes** - existing features unchanged
- New `--estimate-optimal-p` flag (optional, off by default)
- New `--prop-size <size>` flag (only used with optimal-P)
- Step response plots enhanced with optional optimal-P legend
- Console output adds optional optimal-P analysis section

## Documentation

See `OVERVIEW.md` for:
- Theory explanation
- Validation methodology  
- Target values for each prop size
- Physics background (I ∝ r²)

## Known Limitations

- Targets are provisional (need more flight data validation)
- Works best with clean stick inputs (quick flips/rolls)
- Requires D-term data for noise analysis
- Yaw axis not analyzed (intentionally - different dynamics)

## Feedback Wanted

- Do P/D recommendations improve flight performance?
- Are prop-size targets accurate for your setup?
- Is the output clear and actionable?
- Any confusing warnings or terminology?
- PNG legend layout improvements?

---

**Testing this branch helps validate the physics theory with real-world data* 🚁

## Summary by CodeRabbit

* **New Features**
  * Optional Optimal P Estimation: prop-size (1–15") frame-class targets, Td statistics, high-frequency noise metric, and actionable P recommendations with console reports.

* **Plots**
  * Step-response plots can overlay Optimal P analysis summaries and recommendation labels per axis.

* **Documentation**
  * Expanded README and overview: new grouped CLI help, --estimate-optimal-p/--prop-size flags, usage examples, and console output descriptions.

* **Tests**
  * Unit tests added to validate frame-size Td target lookup and expected targets.

### CODE REVIEW

File: src/data_analysis/optimal_p_estimation.rs [hunk 1]
```diff
@@ -0,0 +1,621 @@
...
+
+        // Classify deviation
+        let td_deviation = if td_deviation_percent > 30.0 {
+            TdDeviation::SignificantlySlower
+        } else if td_deviation_percent > 15.0 {
+            TdDeviation::ModeratelySlower
+        } else if td_deviation_percent < -15.0 {
+            TdDeviation::SignificantlyFaster
+        } else {
+            TdDeviation::WithinTarget
```
[the remark this answers is not in the document]
Comment (nerdCopter) [reply src/data_analysis/optimal_p_estimation.rs]: @coderabbitai, resolved?

[the remark this answers is not in the document]
Comment (nerdCopter) [reply src/data_analysis/optimal_p_estimation.rs]: @coderabbitAI, recheck as of commit a5d969a .

My AI Agent reports: 

✅ Imports all constants via [use crate::constants::*;](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) (line 13)
✅ Uses [TD_MEAN_EPSILON](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) (line 198)
✅ Uses [TD_SAMPLES_MIN_FOR_STDDEV](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) (line 204)
✅ Uses [TD_DEVIATION_SIGNIFICANTLY_SLOWER_THRESHOLD](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) (line 285)
✅ Uses [TD_DEVIATION_MODERATELY_SLOWER_THRESHOLD](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) (line 287)
✅ Uses [TD_DEVIATION_SIGNIFICANTLY_FASTER_THRESHOLD](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) (line 289)
The constants are already being used - the refactoring was completed in commit ccecf74. There is no local const MEAN_EPSILON and no hardcoded 30.0, 15.0, -15.0 values.

The file is correct and fully uses the constants from [constants.rs](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html). No changes needed! ✅

**Comment by bw1129** [comment]
looks interesting! Thanks for referencing me

**Comment by nerdCopter** [comment]
Thoughts: 
- I don't like the dependency on empirical thrust-targets table per propeller size. 
- too many variables per aircraft.

What if:
- we calculate thrust of a given aircraft (via Blackbox as that is already how I (AI) calculated my 5-in value).
- i presume we need full-sweep throttle logs. 
- what are an effective X quantity of logs? Can it work with 1? I presume multiple logs would be beneficial but how many are  minimally effective? 

Logistics: 
- how to calculate the thrust given a set of logs? do we iterate each log calculate thrust in totality then reiterate each log for the step response estimated optimal p? Or would we calculate thrust per log and average them? Then perform the step response estimated optimal p? 
- how to logistically wrap this in the code base? Would this also entail grouping Blackbox logs via similar file names (aircraft specific logs)?

@coderabbitAI, review this comment and offer recommendations based on this branch's codebase. I'm willing to fork or re-implement.

**Comment by nerdCopter** [comment]
@coderabbitAI, if you agree i prefer to purge --prop-size and not have it is a fallback.  the empirical 5-in was my 2 different aircraft and in no way should broadly represent all aircraft.  Also the file-name grouping should be intelligent enough to accommodate only grouping similar logs (aircraft name grouping)

File: OVERVIEW.md [hunk 2]
```diff
@@ -189,6 +189,80 @@ The system provides intelligent P:D tuning recommendations based on step-respons
...
+- **Output Format:** Detailed console report with:
+  - Current P gain and measured Td statistics
+  - Frame class comparison and deviation percentage
+  - Physical limit indicators (response speed, noise level, consistency)
+  - Clear recommendation with reasoning
+- **Relationship to P:D Recommendations:**
+  - P:D ratio recommendations (existing feature): Analyze peak overshoot → adjust D-term
+  - Optimal P estimation (new feature): Analyze response timing → adjust P magnitude
+  - Both features are complementary and can run simultaneously for complete tuning guidance
+
```
[the remark this answers is not in the document]
Comment (nerdCopter) [reply OVERVIEW.md:252]: @coderabbitAI, resolved?

File: src/data_analysis/optimal_p_estimation.rs [hunk 3]
```diff
@@ -0,0 +1,698 @@
...
+                write!(f, "{}", message)
+            }
+        }
+    }
+}
+
+impl std::error::Error for AnalysisError {}
+
+/// Minimum valid Td (time to 50%) in milliseconds (domain-appropriate threshold)
+const MIN_TD_MS: f64 = 0.1;
```
[the remark this answers is not in the document]
Comment (nerdCopter) [reply src/data_analysis/optimal_p_estimation.rs]: @coderabbitAI, resolved?

**Comment by nerdCopter** [comment]
as a result I(AI) opened #143 and #141 (141 needs re-basing after 143 potentially merges).

side-note: i'm overusing Claude (usage limits), so i started using Gemini which seems to be okay with math/physics, but is too eager to  agree with human-engineer, has horrible patching-logic, and also fails to follow directions often. Crossing my fingers :crossed_fingers: ... but also was forced to reproduce #143 via Claude-Code from scratch/documentation.

@coderabbitAI, can you investigate #143 on the optimal P/D tuning aspects? then investigate #141 as the solution to our former conversation about "feature/torque-inertia-profiler"

**Comment by nerdCopter** [comment]
@coderabbitai — feedback on your PR #143 observations, items 3 and 4:

**Item 3 — D−1 hint label optionality:** Already handled. The current output reads `Recommendation (conservative): D≈{n} (P={p}) [optional D−1]` — the `[optional D−1]` suffix makes the optionality explicit. Non-issue.

**Item 4 — +2% rounding to no change:** Confirmed real. For small P values (e.g. P=30, D=20, P:D=1.5), `recommended_d = round(30 / 1.47) = 20` — identical to current D, so the Acceptable zone would display `Recommendation (conservative): D≈20` when D is already 20.

Two approaches were considered:
- **Min +1 guard**: clamp `max(rec_d, current_d + 1)` for the Acceptable zone — guarantees a visible actionable nudge.
- **Suppress when identical**: skip storing the recommendation if computed D equals current D.

Decision: leaving as a known limitation for now. The edge case is narrow (only bites when P is small relative to D), and adding the guard introduces zone-specific logic in the computation path that adds complexity disproportionate to the benefit. If it surfaces in real logs it will be revisited.

**Comment by nerdCopter** [comment]
mistakenly merged, then reverted, then force-pushed old master.

**Comment by nerdCopter** [comment]
@coderabbitAI — Following up on the CodeRabbit review at #132 (comment 4501386876) regarding PR #141. #141 has been closed and superseded by #145 (`feature/tip-squash-rebase`), a clean squash-rebase onto master.

Changes made in response to that feedback in #145:

**[Important] BTFL filename grouping fix** — `extract_aircraft_key()` now appends the craft name from the post-date suffix when the prefix ends with `BLACKBOX_LOG`. `BTFL_BLACKBOX_LOG_YYYYMMDD_HHMMSS_CRAFTNAME.NN.csv` now correctly keys to `BTFL_BLACKBOX_LOG_CRAFTNAME` instead of collapsing all files to `BTFL_BLACKBOX_LOG`.

**`TORQUE_PROFILER_SETTLE_SAMPLES`** — Increased from 3 to 5 samples, covering ~5 ms ESC/motor lag at 1 kHz (addresses the concern that 3 ms may be too short for larger/slower aircraft).

**`TORQUE_PROFILER_ACHIEVABILITY_FACTOR`** — Doc comment updated to note empirical calibration on a 5" 6S freestyle build (HELIO H7) and that adjustment may be needed for other aircraft classes.

**Yaw axis** — Added comment to `extract_punch_ratios()` clarifying that axis 2 (Yaw) is collected for diagnostic completeness but is excluded from optimal-P analysis.

**Double-parse TODO** — Noted as a performance concern; not fixed in this PR (correctness is unaffected).

Additionally, two verbiage/plot issues found during rebase review were fixed:
- `"Conservative recommendation:"` / `"Moderate recommendation:"` in `plot_step_response.rs` normalised to `"Recommendation (conservative):"` / `"Recommendation (moderate):"` (matching `main.rs` console output)
- Duplicate `"Recommendation: None (Optimal)"` legend entry removed from PNG when axis is in optimal zone
